[RDY] vim-patch:8.0.{0714, 0720}

gmelodie opened the pull request:
patch 8.0.0714

**Problem:** When a timer causes a command line redraw the " that is displayed for CTRL-R goes missing.
**Solution:** Remember an extra character to display.
https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131

patch 8.0.0720
    
**Problem:** Unfinished mapping not displayed when running timer.
**Solution:** Also use the extra_char while waiting for a mapping and digraph.
https://github.com/vim/vim/commit/6a77d2667e982655f6adacee774ee7aa2581bd8a

Commit:
vim-patch:8.0.0714: when a timer causes a command line redraw " goes missing

Problem:    When a timer causes a command line redraw the " that is displayed
            for CTRL-R goes missing.
Solution:   Remember an extra character to display.
https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -1342,6 +1344,7 @@ static int command_line_handle_key(CommandLineState *s)
 
   case Ctrl_R:                        // insert register
     putcmdline('"', true);
+    extra_char = '"';
```
We already have `ccline.special_char` in nvim for this, but it needs to be made unconditional in `putcmdline`

janlazo commented:
This patch should go together with vim-patch:8.0.0720.

Commit:
Use ccline.special_char instead of extra_char

Commit:
vim-patch:8.0.0720: unfinished mapping not displayed when running timer

Problem:    Unfinished mapping not displayed when running timer.
Solution:   Also use the extra_char while waiting for a mapping and digraph.
            (closes vim/vim#1844)
https://github.com/vim/vim/commit/6a77d2667e982655f6adacee774ee7aa2581bd8a

Commit:
Merge branch 'vim-8.0.0714' of github.com:LTKills/neovim into vim-8.0.0720

Commit:
Fix CI errors

justinmk commented:
cmdline_speci is failing, maybe it needs to be updated? You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/ui/cmdline_spec.lua make functionaltest

Commit:
Merge branch 'master' of https://github.com/neovim/neovim into vim-8.0.0714

LTKills commented:
The failing test is:
```
-- erase information, so we check if it is retransmitted
    command("mode")
    screen:expect{grid=[[
      ^                         |
      {1:~                        }|
      {1:~                        }|
      {1:~                        }|
                               |
    ]], cmdline={{
      firstc = ":",
      content = {{"xx"}},
      pos = 2,
      special = {'"', true}
    }, {
      firstc = "=",
      content = {{"1"}, {"+"}, {"2"}},
      pos = 3,
    }}, reset=true}
```

What is actually being returned is this:

```
cmdline={{
      firstc = ":",
      content = {{"xx"}},
      pos = 2,
      <missing special character here>
    }
```

It seems the `mode` command redraws the screen, I'm not sure if the `<C-R>`should be present after the screen is redrawn.

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -1351,7 +1351,8 @@ static int command_line_handle_key(CommandLineState *s)
       s->c = plain_vgetc();              // CTRL-R CTRL-R <char>
     }
-    --no_mapping;
+    ccline.special_char = NUL;
```
This is incorrect, as it is scoped in nvim (unlike the original code). This should fix the commented issue also.

Commit:
Fix failing tests

Commit:
Fix persisting test error

Commit:
Merge branch 'vim-8.0.0714' of github.com:LTKills/neovim into vim-8.0.0714

Commit:
Parser highlighting does not override special character

justinmk commented:
LGTM. is this ready?

justinmk commented on src/nvim/ex_getln.c:
```diff
@@ -1447,6 +1447,9 @@ static int command_line_handle_key(CommandLineState *s)
     set_cmdspos_cursor();
+    if (ccline.special_char != NUL) {
+      putcmdline(ccline.special_char, ccline.special_shift);
+    }
```
@LTKills  Why do this here, in the `K_LEFT`/`K_S_LEFT`/... case? I don't see anything like that in https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131 .

justinmk commented:
merged. https://github.com/neovim/neovim/commit/3b7a4f233bd53eb901904992227be3fffe5e621a
